modules/dnn/src/layers/reduce_layer.cpp

```diff
@@ -0,0 +1,436 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "../precomp.hpp"
+#include "opencv2/core/hal/intrin.hpp"
+#include "../op_cuda.hpp"
+#include "../op_webnn.hpp"
+
+#include <float.h>
+#include <algorithm>
+#include <numeric>
+using std::max;
+using std::min;
+
+#include <opencv2/core/utils/logger.hpp>
+
+namespace cv
+{
+namespace dnn
+{
+
+class ReduceLayerImpl CV_FINAL : public ReduceLayer
+{
+public:
+    ReduceLayerImpl(const LayerParams& params)
+    {
+        // set reduce type
+        CV_Assert(params.has("reduce"));
+        String reduceType = toLowerCase(params.get<String>("reduce"));
+        if (reduceType == "max")
+            type = MAX;
+        else if (reduceType == "min")
+            type = MIN;
+        else if (reduceType == "ave")
+            type = AVE;
+        else if (reduceType == "sum")
+            type = SUM;
+        else if (reduceType == "sum_square")
+            type = SUM_SQUARE;
+        else if (reduceType == "l1")
+            type = L1;
+        else if (reduceType == "l2")
+            type = L2;
+        else if (reduceType == "log_sum")
+            type = LOG_SUM;
+        else if (reduceType == "log_sum_exp")
+            type = LOG_SUM_EXP;
+        else if (reduceType == "prod")
+            type = PROD;
+        else
+            CV_Error(Error::StsBadArg, "Unknown pooling type \"" + reduceType + "\"");
+
+        // set deleted dims
+        CV_Assert(params.has("deleted_dims"));
+        DictValue tempDims = params.get("deleted_dims");
+        int i, n = tempDims.size();
+        deletedDims.resize(n);
+        for (i = 0; i < n; i++)
+        {
+            deletedDims[i] = tempDims.get<int>(i);
+        }
+    }
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        if (backendId == DNN_BACKEND_OPENCV)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    class ReduceOpBase
+    {
+    public:
+        bool hasSIMD = false;
+        ReduceOpBase() {}
+        virtual ~ReduceOpBase() {};
+        virtual float apply(const float*, const float*, const float ikarea = 1.0f) = 0;
```

Comment:
Please drop inheritance from functors: https://github.com/alalek/opencv/commit/pr21601_r

(with similar change for int8 implementation)